src/cachemachine/cachechecker.py

```diff
@@ -0,0 +1,70 @@
+from collections import defaultdict
+from typing import Dict, List, Set
+
+import structlog
+from kubernetes.client.api import core_v1_api
+
+logger = structlog.get_logger(__name__)
+
+
+class CacheChecker:
+    common_cache: Set[str]
+    labels: Dict[str, str]
+    nodes_exist: bool
+    recommended_names: Dict[str, Set[str]]
+    recommended_image_urls: List[str]
+
+    def __init__(
+        self, labels: Dict[str, str], recommended_image_urls: List[str]
+    ):
+        self.api = core_v1_api.CoreV1Api()
+        self.labels = labels
+        self.recommended_image_urls = recommended_image_urls
+        self._reset()
+
+    def _reset(self) -> None:
+        self.common_cache = set()
+        self.nodes_exist = False
+        self.recommended_names = defaultdict(set)
+
+    def check(self) -> None:
+        self._reset()
+
+        nodes = self.api.list_node().items
+        logger.debug(f"Inspecting {nodes}")
+
+        for n in nodes:
+            node_name = n.metadata.name
+
+            # This is a bit tricky.  The images is a list,
+            # each item containing a particular image, and containing
+            # a list of all the names it is known by.
+            images = []
+            recommended_names: Dict[str, Set[str]] = defaultdict(set)
+            for i in n.status.images:
+                for url in self.recommended_image_urls:
+                    if url in i.names:
+                        recommended_names[url] |= set(i.names)
+
+                images.extend(i.names)
+
+            logger.debug(f"{node_name} images: {images}")
+            logger.debug(f"{node_name} recommended: {recommended_names}")
+
+            # If the labels we have are a subset of the labels
+            # of the node, include those images.
+            logger.debug(f"{node_name} labels: {n.metadata.labels}")
+            if self.labels.items() <= n.metadata.labels.items():
+                if not self.nodes_exist:
+                    # For the first node we find, that set is the initial
+                    # common set.
+                    self.nodes_exist = True
+                    self.common_cache = set(images)
+                    self.recommended_names = recommended_names
+                else:
+                    # Take the intersection of all the images on all the
+                    # nodes so far.
+                    self.common_cache &= set(images)
+
+                    for url in self.recommended_image_urls:
+                        self.recommended_names[url] &= recommended_names[url]
```

Comment:
Recommended names are the only names.  And by names, this means tags.  The reason for the intersection is that on different nodes, depending on what they've pulled, recommended might point to different things.  On one node, the recommended hash and image might point to weekly n, and another one it might point to weekly n+1.  So I only keep track of the tags if they are consistent across the cluster.  Will add a comment to this effect, the parts about recommended are the most annoying, and I'm mostly handling it because it's part of what we did before (although I'm still not sure we need to know all this).

Reply:
I'm confused why the recommended image would be inconsistent across the cluster. Isn't the recommended image the one that Nublado is going to use by default, and therefore all the nodes in the cluster should prepull the same image, matching the one that Nublado is going to launch?

Also, I don't understand what these URLs are and what they point to and why they have names, but I think this is just another way of saying that I need to read the documentation you're working on. :)

Reply:
> Isn't the recommended image the one that Nublado is going to use by default, and therefore all the nodes in the cluster should prepull the same image, matching the one that Nublado is going to launch?

Yes this is true, but the recommended image and what it's pointing to is only updated after it's pulled.  Otherwise it'll show the hash of the latest recommended that was pulled, so sometimes you can have an inconsistent state where recommended may point to one hash on one kubernetes node, but a different hash on a different node.  Once we pull recommended on all the nodes though, they should be consistent.